Add SpanJson to Benchmarks

Tornhoof opened the pull request:
- Added Reference to nuget package.
- Added Utf16 Version to Serializer and Deserializer Comparison. I postfixed the name with Utf16 as this is consistent with my own tests and it's easier to add an Utf8 version later.
- Added both Utf8/Utf16 versions to JsonReaderComparison.

Benchmark results are in #2288 and below
<details>

**JsonReaderComparison**
``` ini

BenchmarkDotNet=v0.10.14.534-nightly, OS=Windows 10.0.17134
Intel Core i7-4790K CPU 4.00GHz (Haswell), 1 CPU, 8 logical and 4 physical cores
Frequency=3906248 Hz, Resolution=256.0001 ns, Timer=TSC
.NET Core SDK=2.1.300-rtm-008823
  [Host]     : .NET Core 2.1.0-rtm-26509-04 (CoreCLR 4.6.26509.03, CoreFX 4.6.26509.03), 64bit RyuJIT
  DefaultJob : .NET Core 2.1.0-rtm-26509-04 (CoreCLR 4.6.26509.03, CoreFX 4.6.26509.03), 64bit RyuJIT


```
|              Method |        Target |  JsonStringName |         Mean |       Error |      StdDev | Scaled | ScaledSD |  Gen 0 | Allocated |
|-------------------- |-------------- |---------------- |-------------:|------------:|------------:|-------:|---------:|-------:|----------:|
|    **ReaderNewtonsoft** | **InvariantUtf8** | **HeavyNestedJson** | **17,399.87 ns** |  **29.3604 ns** |  **24.5173 ns** |   **1.00** |     **0.00** | **2.1057** |    **8904 B** |
|     ReaderCorefxlab | InvariantUtf8 | HeavyNestedJson |  5,501.33 ns |   7.4602 ns |   6.9782 ns |   0.32 |     0.00 |      - |       0 B |
|      ReaderUtf8Json | InvariantUtf8 | HeavyNestedJson | 11,634.98 ns |  17.9349 ns |  16.7763 ns |   0.67 |     0.00 |      - |       0 B |
|       ReaderJayrock | InvariantUtf8 | HeavyNestedJson | 41,562.38 ns | 786.3366 ns | 772.2876 ns |   2.39 |     0.04 | 4.9438 |   20880 B |
|       ReaderLitJson | InvariantUtf8 | HeavyNestedJson | 54,013.30 ns |  46.0975 ns |  43.1196 ns |   3.10 |     0.00 | 1.6479 |    6928 B |
|  ReaderSpanJsonUtf8 | InvariantUtf8 | HeavyNestedJson |  8,890.21 ns |  64.5886 ns |  60.4162 ns |   0.51 |     0.00 |      - |       0 B |
|                     |               |                 |              |             |             |        |          |        |           |
|    **ReaderNewtonsoft** | **InvariantUtf8** |      **HelloWorld** |    **378.29 ns** |   **0.4317 ns** |   **0.4038 ns** |   **1.00** |     **0.00** | **0.5641** |    **2368 B** |
|     ReaderCorefxlab | InvariantUtf8 |      HelloWorld |     64.77 ns |   0.1076 ns |   0.1006 ns |   0.17 |     0.00 |      - |       0 B |
|      ReaderUtf8Json | InvariantUtf8 |      HelloWorld |     82.64 ns |   0.0954 ns |   0.0892 ns |   0.22 |     0.00 |      - |       0 B |
|       ReaderJayrock | InvariantUtf8 |      HelloWorld |    553.50 ns |   4.2478 ns |   3.9734 ns |   1.46 |     0.01 | 0.3405 |    1432 B |
|       ReaderLitJson | InvariantUtf8 |      HelloWorld |    792.08 ns |   0.7162 ns |   0.6349 ns |   2.09 |     0.00 | 0.1764 |     744 B |
|  ReaderSpanJsonUtf8 | InvariantUtf8 |      HelloWorld |     64.87 ns |   0.2278 ns |   0.2019 ns |   0.17 |     0.00 |      - |       0 B |

**JsonDeserializerComparison_FromString_LoginViewModel**

``` ini

BenchmarkDotNet=v0.10.14.534-nightly, OS=Windows 10.0.17134
Intel Core i7-4790K CPU 4.00GHz (Haswell), 1 CPU, 8 logical and 4 physical cores
Frequency=3906248 Hz, Resolution=256.0001 ns, Timer=TSC
.NET Core SDK=2.1.300-rtm-008823
  [Host]     : .NET Core 2.1.0-rtm-26509-04 (CoreCLR 4.6.26509.03, CoreFX 4.6.26509.03), 64bit RyuJIT
  DefaultJob : .NET Core 2.1.0-rtm-26509-04 (CoreCLR 4.6.26509.03, CoreFX 4.6.26509.03), 64bit RyuJIT


```
|        Method |         Mean |         Error |        StdDev | Scaled | ScaledSD |  Gen 0 |  Gen 1 |  Gen 2 | Allocated |
|-------------- |-------------:|--------------:|--------------:|-------:|---------:|-------:|-------:|-------:|----------:|
|    Newtonsoft |   1,200.3 ns |     4.8153 ns |     4.5042 ns |   1.00 |     0.00 | 0.6618 |      - |      - |    2784 B |
|           Jil |     346.5 ns |     1.1224 ns |     0.9950 ns |   0.29 |     0.00 | 0.0644 |      - |      - |     272 B |
|      Utf8Json |     469.2 ns |     1.2537 ns |     1.1114 ns |   0.39 |     0.00 | 0.0682 |      - |      - |     288 B |
|        YSharp | 360,634.0 ns | 3,711.9044 ns | 3,472.1174 ns | 300.46 |     3.00 | 9.2773 | 4.3945 | 0.4883 |   38578 B |
|      FastJson |   2,172.7 ns |     5.8238 ns |     5.4476 ns |   1.81 |     0.01 | 0.5722 |      - |      - |    2416 B |
| SpanJsonUtf16 |     200.1 ns |     0.6457 ns |     0.6040 ns |   0.17 |     0.00 | 0.0417 |      - |      - |     176 B |

**JsonSerializerComparison_ToString_LoginViewModel**

``` ini

BenchmarkDotNet=v0.10.14.534-nightly, OS=Windows 10.0.17134
Intel Core i7-4790K CPU 4.00GHz (Haswell), 1 CPU, 8 logical and 4 physical cores
Frequency=3906248 Hz, Resolution=256.0001 ns, Timer=TSC
.NET Core SDK=2.1.300-rtm-008823
  [Host]     : .NET Core 2.1.0-rtm-26509-04 (CoreCLR 4.6.26509.03, CoreFX 4.6.26509.03), 64bit RyuJIT
  DefaultJob : .NET Core 2.1.0-rtm-26509-04 (CoreCLR 4.6.26509.03, CoreFX 4.6.26509.03), 64bit RyuJIT


```
|        Method |       Mean |      Error |     StdDev | Scaled | ScaledSD |  Gen 0 | Allocated |
|-------------- |-----------:|-----------:|-----------:|-------:|---------:|-------:|----------:|
|    Newtonsoft |   751.3 ns | 12.8312 ns | 11.3745 ns |   1.00 |     0.00 | 0.3605 |    1512 B |
|           Jil |   364.3 ns |  6.4239 ns |  5.3642 ns |   0.48 |     0.01 | 0.1750 |     736 B |
|      Utf8Json |   216.6 ns |  0.1600 ns |  0.1157 ns |   0.29 |     0.00 | 0.0455 |     192 B |
|      FastJson | 1,168.1 ns |  1.5533 ns |  1.3769 ns |   1.55 |     0.02 | 0.7019 |    2952 B |
|       Manatee | 4,403.8 ns |  6.8871 ns |  6.4422 ns |   5.86 |     0.08 | 1.0147 |    4272 B |
| SpanJsonUtf16 |   193.5 ns |  0.1432 ns |  0.1196 ns |   0.26 |     0.00 | 0.0455 |     192 B |

</details>

Tornhoof commented:
Tagging @ahsonkhan as requested.

ahsonkhan commented on tests/Benchmarks/System.Text.JsonLab/Comparison/JsonReader/JsonReaderComparison.cs:
```diff
@@ -88,5 +88,27 @@ public void ReaderLitJson()
+        [Benchmark]
+        public void ReaderSpanJsonUtf16()
+        {
+            SpanJson.JsonReader<char> json = new SpanJson.JsonReader<char>(_str);
```
Can you create the reader in GlobalSetup instead? Also, we use the EncoderTarget to enumerate the encoding (currently it only has UTF-8, but if needed, we can add the UTF-16 as well). So, having separate benchmarks like this won't work (since we wuold end up running ReaderSpanJsonUtf16 twice (once for UTF-8 and once for UTF-16) and same with the other benchmark.

Is there a way to parameterize the encoding out and set it once so we can have only one benchmark? If not, I would only add the UTf8 one for now.

Tornhoof replied:
I think i'll remove the Utf16 version, it's not possible to parametrize the encoding out, as neither my API supports it nor can I add a ref struct as a field in that class.

@adamsitnik Benchmark.NET does not support having generic arguments to benchmark methods at the moment, right?
Something like:
```csharp
[ParamsSource(nameof(TypeMap))]
public Type TParam;
public Type TypeMap()
{
    switch (Target)
    {
        case EncoderTarget.InvariantUtf16:
            return typeof(char);
        case EncoderTarget.InvariantUtf8:
            return typeof(byte);
    }
}
[Benchmark]
public void MyBenchmark<TParam>() where TParam: struct
{
  // still have to solve the different arguments (_str vs. _data) then.
}
```
having some switch with cases for ``EncoderTarget`` would probably influence the benchmark too much. 
In any case this benchmark is not flattering for SpanJson, apparently having whitespace in jsons makes it 2-3 slower than without whitespace (utf8json has the same problem)

adamsitnik replied:
@Tornhoof BDN supports generic types with benchmarks, but not non-generic types with generic benchmark methods

However, you can add this requirement here: https://github.com/dotnet/BenchmarkDotNet/issues/745 I can't promise to add it any time soon, but I might add it to my TODO list

Tornhoof replied:
@adamsitnik I don't think that's necessary, my example above is a highly specific use case to get around benchmark design and limitations in SpanJson.

Tornhoof commented on tests/Benchmarks/System.Text.JsonLab/Comparison/JsonDeserializer/JsonDeserializerComparison_FromString.cs:
```diff
@@ -92,5 +95,13 @@ public T Manatee_()
             return deserialized;
         }
+
+        [Benchmark(Description = "SpanJsonUtf16")]
```
Ah ok, I'll change that.

Tornhoof replied:
I've removed those descriptions from other json benchmarks too, where the rules above don't apply

KrzysztofCwalina commented:
It would be good if the description of this PR included the pref results (i.e. BenchmarkDotNet report)

Tornhoof commented:
I added the benchmark results to the description too and added the actual benchmark names (those I missed before).
@adamsitnik  Thanks for the praise, but most of that belongs to all of you implementing the performant Span and related APIs.

KrzysztofCwalina commented:
@tornhoof, Thanks!

@ahsonkhan, @joshfree, we need to think which set of technologies are included in this benchmark suite. So that we don't end up with 100s of libraries being benchmarked here. I would propose something like the following:

1. Corfxlab tenchnologies ... duh :-)
2. Popular established technologies (JSON.NET, JIl) we chose.
3. The fastest technology submitted by the technology owner.

joshfree commented:
@KrzysztofCwalina agreed.  Also we should be clear in the benchmarks that this is temporary/experimental for @ahsonkhan 's investigation - and at some point in the near future the benchmarks will go away.
